src/flowfield.cpp

```diff
@@ -0,0 +1,811 @@
+	std::unique_ptr<CostField>(new CostField()),
+	std::unique_ptr<CostField>(new CostField()),
+	std::unique_ptr<CostField>(new CostField()),
+};
+
+// Flow field cache for ground, hover and lift movement types
+std::array<std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>, 4> flowfieldCaches {
+	std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>(new std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>()),
+	std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>(new std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>()),
+	std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>(new std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>()),
+	std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>(new std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>())
+};
+
+bool tryGetFlowfieldForTarget(unsigned int targetX, unsigned int targetY, PROPULSION_TYPE propulsion, unsigned int &flowfieldId){
+	const auto flowfieldCache = flowfieldCaches[propulsionToIndex.at(propulsion)].get();
+
+	std::vector<ComparableVector2i> goals { { map_coord(targetX), map_coord(targetY) } };
+
+ 	// this check is already done in fpath.cpp.
+	// TODO: we should perhaps refresh the flowfield instead of just bailing here.
+	if (!flowfieldCache->count(goals)) {
+		return false;
+	}
+
+	const auto flowfield = flowfieldCache->at(goals).get();
+
+	flowfieldId = flowfield->id;
+
+	return true;
+}
+
+std::map<unsigned int, Flowfield*> flowfieldById;
+bool tryGetFlowfieldVector(unsigned int flowfieldId, int x, int y, Vector2f& vector){
+	if(!flowfieldById.count(flowfieldId)){
+		return false;
+	}
+
+	auto flowfield = flowfieldById.at(flowfieldId);
+
+	auto v = flowfield->getVector(x, y);
+	vector = { v.x, v.y };
+
+	return true;
+}
+
+struct Node {
+	unsigned short predecessorCost;
+	unsigned int index;
+
+	bool operator<(const Node& other) const {
+		// We want top element to have lowest cost
+		return predecessorCost > other.predecessorCost;
+	}
+};
+
+void calculateIntegrationField(const std::vector<ComparableVector2i>& points, IntegrationField* integrationField, CostField* costField);
+void calculateFlowfield(Flowfield* flowField, IntegrationField* integrationField);
+
+void processFlowfield(FLOWFIELDREQUEST request) {
+
+	// NOTE for us noobs!!!! This function is executed on its own thread!!!!
+
+	const auto flowfieldCache = flowfieldCaches[propulsionToIndex.at(request.propulsion)].get();
+	const auto costField = costFields[propulsionToIndex.at(request.propulsion)].get();
+
+	std::vector<ComparableVector2i> goals { request.goal }; // TODO: multiple goals for formations
+
+ 	// this check is already done in fpath.cpp.
+	// TODO: we should perhaps refresh the flowfield instead of just bailing here.
+	if (flowfieldCache->count(goals)) {
```

Comment:
I think all uses of `flowfieldCache` should be under a mutex. Otherwise compiler will assume single-thread behavior and might optimize in unexpected ways.